File: apps/android/journal/SurveyGuruJournal.swift
// Appended to the pinned GPS plugin by apply-journal.cjs.
private final class SurveyGuruJournal {
    static let shared = SurveyGuruJournal()
    private let lock = NSLock()
    private var owner: String?
    private var session: String?
    private var lastTime: TimeInterval = 0
    private var failure: Error?
    func configure(owner: String, session: String) {
        lock.lock(); defer { lock.unlock() }
        self.owner = owner; self.session = session; lastTime = 0; failure = nil
    }
    func stop() { lock.lock(); defer { lock.unlock() }; owner = nil; session = nil }
    private func directory() throws -> URL {
        var url = try FileManager.default.url(for: .applicationSupportDirectory, in: .userDomainMask, appropriateFor: nil, create: true).appendingPathComponent("SurveyGuruLocations", isDirectory: true)
        try FileManager.default.createDirectory(at: url, withIntermediateDirectories: true)
        var values = URLResourceValues(); values.isExcludedFromBackup = true; try url.setResourceValues(values)
        return url
    }
    func append(_ location: CLLocation) {
        lock.lock(); defer { lock.unlock() }
        guard let owner = owner, let session = session, location.horizontalAccuracy >= 0, location.horizontalAccuracy <= 100,
              location.timestamp.timeIntervalSince1970 - lastTime >= 10 else { return }
        if #available(iOS 15, *), location.sourceInformation?.isSimulatedBySoftware == true { return }
        do {
            let dir = try directory()
            let files = try FileManager.default.contentsOfDirectory(at: dir, includingPropertiesForKeys: nil)
            if files.count >= 65000 { throw NSError(domain: "SurveyGuru", code: 1, userInfo: [NSLocalizedDescriptionKey: "Location storage is full. Synchronise before continuing."]) }
            let id = UUID().uuidString.lowercased()
            let formatter = ISO8601DateFormatter(); formatter.formatOptions = [.withInternetDateTime, .withFractionalSeconds]
            let point: [String: Any] = ["eventId": id, "ownerId": owner, "sessionId": session, "capturedAt": formatter.string(from: location.timestamp), "latitude": location.coordinate.latitude, "longitude": location.coordinate.longitude, "accuracyMetres": location.horizontalAccuracy, "source": "native_background"]
            let data = try JSONSerialization.data(withJSONObject: point)
            try data.write(to: dir.appendingPathComponent(id + ".json"), options: [.atomic, .completeFileProtectionUntilFirstUserAuthentication])
            lastTime = location.timestamp.timeIntervalSince1970
        } catch { failure = error; self.owner = nil; self.session = nil }
    }
    func pending(owner: String, session: String) throws -> [[String: Any]] {
        lock.lock(); defer { lock.unlock() }
        if let failure = failure { throw failure }
        let files = try FileManager.default.contentsOfDirectory(at: directory(), includingPropertiesForKeys: nil)
        var points: [[String: Any]] = []
        for file in files where file.pathExtension == "json" {
            if let point = try JSONSerialization.jsonObject(with: Data(contentsOf: file)) as? [String: Any], point["ownerId"] as? String == owner, point["sessionId"] as? String == session { points.append(point) }
        }
        return Array(points.sorted { ($0["capturedAt"] as? String ?? "") < ($1["capturedAt"] as? String ?? "") }.prefix(200))
    }
    func acknowledge(owner: String, session: String, eventId: String) throws {
        lock.lock(); defer { lock.unlock() }
        guard UUID(uuidString: eventId) != nil else { throw NSError(domain: "SurveyGuru", code: 2) }
        let file = try directory().appendingPathComponent(eventId.lowercased() + ".json")
        if !FileManager.default.fileExists(atPath: file.path) { return }
        guard let point = try JSONSerialization.jsonObject(with: Data(contentsOf: file)) as? [String: Any], point["ownerId"] as? String == owner, point["sessionId"] as? String == session else { throw NSError(domain: "SurveyGuru", code: 3) }
        try FileManager.default.removeItem(at: file)
    }
}

File: apps/android/journal/SurveyGuruJournal.java
package com.capgo.capacitor_background_geolocation;

import android.content.Context;
import android.location.Location;
import androidx.core.location.LocationCompat;
import org.json.JSONArray;
import org.json.JSONObject;
import java.io.File;
import java.io.FileOutputStream;
import java.nio.charset.StandardCharsets;
import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.io.FileInputStream;
import java.text.SimpleDateFormat;
import java.util.Arrays;
import java.util.Date;
import java.util.Locale;
import java.util.TimeZone;
import java.util.UUID;

/** Private native disk journal: write before emitting a WebView callback, delete only after server acknowledgement. */
final class SurveyGuruJournal {
    private static String owner;
    private static String session;
    private static String failure;
    private static long lastTime;
    static synchronized void configure(String nextOwner, String nextSession) {
        owner = nextOwner; session = nextSession; lastTime = 0; failure = null;
    }
    static synchronized void stop() { owner = null; session = null; }
    private static File directory(Context context) throws Exception {
        File dir = new File(context.getNoBackupFilesDir(), "survey-guru-locations");
        if (!dir.exists() && !dir.mkdirs()) throw new Exception("Cannot create location storage.");
        return dir;
    }
    private static String read(File file) throws Exception {
        StringBuilder value = new StringBuilder();
        try (BufferedReader reader = new BufferedReader(new InputStreamReader(new FileInputStream(file), StandardCharsets.UTF_8))) {
            String line; while ((line = reader.readLine()) != null) value.append(line);
        }
        return value.toString();
    }
    static synchronized void append(Context context, Location location) {
        if (owner == null || session == null || LocationCompat.isMock(location) || location.getTime() <= lastTime || location.getTime() - lastTime < 10000 || !location.hasAccuracy() || location.getAccuracy() < 0 || location.getAccuracy() > 100) return;
        try {
            File dir = directory(context);
            String[] names = dir.list();
            if (names != null && names.length >= 65000) throw new Exception("Location storage is full. Synchronise before continuing.");
            String id = UUID.randomUUID().toString();
            SimpleDateFormat date = new SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ss.SSS'Z'", Locale.US); date.setTimeZone(TimeZone.getTimeZone("UTC"));
            JSONObject point = new JSONObject().put("eventId", id).put("ownerId", owner).put("sessionId", session)
                .put("capturedAt", date.format(new Date(location.getTime()))).put("latitude", location.getLatitude()).put("longitude", location.getLongitude())
                .put("accuracyMetres", location.getAccuracy()).put("source", "native_background");
            File temp = new File(dir, id + ".tmp");
            try (FileOutputStream output = new FileOutputStream(temp)) { output.write(point.toString().getBytes(StandardCharsets.UTF_8)); output.getFD().sync(); }
            if (!temp.renameTo(new File(dir, id + ".json"))) throw new Exception("Could not commit location to disk.");
            lastTime = location.getTime();
        } catch (Exception error) { failure = error.getMessage(); owner = null; session = null; }
    }
    static synchronized JSONArray pending(Context context, String userId, String sessionId) throws Exception {
        if (failure != null) throw new Exception(failure);
        File[] files = directory(context).listFiles((dir, name) -> name.endsWith(".json"));
        JSONArray result = new JSONArray();
        if (files == null) return result;
        Arrays.sort(files, (a,b) -> Long.compare(a.lastModified(), b.lastModified()));
        for (File file : files) {
            JSONObject point = new JSONObject(read(file));
            if (userId.equals(point.optString("ownerId")) && sessionId.equals(point.optString("sessionId"))) { result.put(point); if (result.length() >= 200) break; }
        }
        return result;
    }
    static synchronized void acknowledge(Context context, String userId, String sessionId, String eventId) throws Exception {
        if (eventId == null || !eventId.matches("[a-f0-9-]{36}")) throw new Exception("Invalid event ID.");
        File file = new File(directory(context), eventId + ".json");
        if (!file.exists()) return;
        JSONObject point = new JSONObject(read(file));
        if (!userId.equals(point.optString("ownerId")) || !sessionId.equals(point.optString("sessionId"))) throw new Exception("Location belongs to another account or session.");
        if (!file.delete()) throw new Exception("Could not acknowledge saved location.");
    }
}
